pkg/virt-controller/watch/export/export.go

```diff
@@ -384,19 +459,142 @@ func (ctrl *VMExportController) handleIsSourcePvc(vmExport *exportv1.VirtualMach
+		if err := ctrl.handlePodSucceededOrFailed(vmExport, pod); err != nil {
+			return 0, err
+		}
+	}
+
+	return ctrl.updateVMExportPvcStatus(vmExport, pvcs, pod, service, inUse, isPopulated)
+
+}
+
+func (ctrl *VMExportController) handleIsVmSnapshot(vmExport *exportv1.VirtualMachineExport, service *corev1.Service) (time.Duration, error) {
+	var pvcs []*corev1.PersistentVolumeClaim
+	vmSnapshot, exists, err := ctrl.getVmSnapshot(vmExport.Namespace, vmExport.Spec.Source.Name)
+	restoreableSnapshots := 0
+	if err != nil {
+		return 0, err
+	} else if exists {
+		if vmSnapshot.Status.ReadyToUse != nil && *vmSnapshot.Status.ReadyToUse {
+			pvcs, restoreableSnapshots, err = ctrl.handlePVCsForVirtualMachineSnapshot(vmExport, vmSnapshot)
+			if err != nil {
 				return 0, err
-			} else {
-				return retry, nil
 			}
 		}
+	} // If snapshot doesn't exist or is not ready, it will fall through to updateVMExportVMSnapshotStatus because no pod will exist and no pvcs
+
+	pod, exists, err := ctrl.getExporterPod(vmExport)
+	if err != nil {
+		return 0, err
+	} else if !exists {
+		if len(pvcs) == restoreableSnapshots && restoreableSnapshots > 0 {
+			pod, err = ctrl.createExporterPod(vmExport, pvcs)
+			if err != nil {
+				return 0, err
+			} else if pod == nil {
+				return 0, nil
+			}
+
+			if err := ctrl.getOrCreateCertSecret(vmExport, pod); err != nil {
+				return 0, err
+			}
+		}
+	} else {
+		if err := ctrl.handlePodSucceededOrFailed(vmExport, pod); err != nil {
+			return 0, err
+		}
+	}
+	return 0, ctrl.updateVMExporVMSnapshotStatus(vmExport, pvcs, pod, service, false)
+}
+
+func (ctrl *VMExportController) handlePodSucceededOrFailed(vmExport *exportv1.VirtualMachineExport, pod *corev1.Pod) error {
+	if pod.Status.Phase == corev1.PodSucceeded || pod.Status.Phase == corev1.PodFailed {
+		// The server died or completed, delete the pod.
+		ctrl.Recorder.Eventf(vmExport, corev1.EventTypeWarning, exporterPodFailedOrCompletedEvent, "Exporter pod %s/%s is in phase %s", pod.Namespace, pod.Name, &pod.Status.Phase)
+		return ctrl.Client.CoreV1().Pods(vmExport.Namespace).Delete(context.Background(), pod.Name, metav1.DeleteOptions{})
 	}
+	return nil
+}
 
-	return ctrl.updateVMExportPvcStatus(vmExport, pvcs, pod, service, inUse)
+func (ctrl *VMExportController) handlePVCsForVirtualMachineSnapshot(vmExport *exportv1.VirtualMachineExport, vmSnapshot *snapshotv1.VirtualMachineSnapshot) ([]*corev1.PersistentVolumeClaim, int, error) {
+	var content *snapshotv1.VirtualMachineSnapshotContent
+	var err error
+	var pvcs []*corev1.PersistentVolumeClaim
+	exists := false
+	totalVolumes := 0
 
+	if vmSnapshot.Status.VirtualMachineSnapshotContentName != nil && *vmSnapshot.Status.VirtualMachineSnapshotContentName != "" {
+		content, exists, err = ctrl.getVmSnapshotContent(vmSnapshot.Namespace, *vmSnapshot.Status.VirtualMachineSnapshotContentName)
+		if err != nil {
+			return nil, 0, err
+		} else if exists {
```

Comment:
There are a LOT of occurrences of this pattern in this file and it is not idiomatic go. 

```golang
if cond1 {
  cond1
} else if cond2 {
   ...
}
``` 

Should be:

```golang
if cond1 {
  return
}

if cond2 {
  ...
}
```